syntax = "proto3";

package services;

// ======================
// ServiÃ§o A - Produtos
// ======================
service ServiceA {
  rpc ListarProdutos (Empty) returns (ListaProdutosResponse);
  rpc ObterProduto (ProdutoRequest) returns (ProdutoResponse);
  rpc CriarProduto (CriarProdutoRequest) returns (ProdutoResponse);
  rpc EditarProduto (ProdutoUpdateRequest) returns (ProdutoResponse); // ðŸ†•
  rpc DeletarProduto (ProdutoRequest) returns (DeleteResponse);       // ðŸ†•
}

// ======================
// ServiÃ§o B - Estoque
// ======================
service ServiceB {
  rpc ObterEstoque (EstoqueRequest) returns (EstoqueResponse);
}

// ======================
// Mensagens compartilhadas
// ======================

message Empty {}

message ProdutoRequest {
  string id = 1;
}

message ProdutoResponse {
  string id = 1;
  string nome = 2;
  string descricao = 3;
  double preco = 4;
}

message CriarProdutoRequest {
  string nome = 1;
  string descricao = 2;
  double preco = 3;
}

message ListaProdutosResponse {
  repeated ProdutoResponse produtos = 1;
}

message ProdutoUpdateRequest {
  string id = 1;
  string nome = 2;
  string descricao = 3;
  double preco = 4;
}

message DeleteResponse {
  bool sucesso = 1;
  string mensagem = 2;
}

message EstoqueRequest {
  string produto_id = 1;
}

message EstoqueResponse {
  string produto_id = 1;
  int32 quantidade = 2;
  string localizacao = 3;
}

// Mensagens genÃ©ricas usadas nos testes
message Request {
  string input = 1;
}

message Response {
  string output = 1;
}
